File: test_bench.ts
// test_bench.ts - Comprehensive benchmark for scheduler testing
import {
  scheduleCallback,
  setTestSchedulerOptions,
  resetSchedulerState,
  getTestSchedulerOptions,
} from "./test_scheduler.js";

// Test parameters
const SLEEP_VALUES = [0, 0.001, 0.002, 0.016]; // in seconds (0, 1ms, 2ms, 16ms)
const THRESHOLDS = [0, 1, 2, 3, 16]; // timeout <= k
const ITERATIONS = 100; // reduced from 400 for faster testing

// Nested test parameters
const NEST_DEPTH = 10; // 10 calls to observe 4ms clamping effect (nestingLevel >= 5 from 6th call)

// High-frequency test parameters
const BURST_SIZE = 100; // number of concurrent callbacks

// Performance measurement
function now(): number {
  if (typeof performance !== "undefined" && performance.now) {
    return performance.now();
  }
  return Date.now();
}

// ============ Helper Functions ============

/**
 * Break timer nesting by yielding to event loop via non-timer task
 * This prevents unintended nested timer clamping
 */
function yieldToEventLoop(): Promise<void> {
  return new Promise((resolve) => {
    if (typeof (globalThis as any).setImmediate === "function") {
      (globalThis as any).setImmediate(resolve);
    } else if (typeof MessageChannel === "function") {
      const channel = new MessageChannel();
      channel.port1.onmessage = () => {
        channel.port1.close();
        channel.port2.close();
        resolve();
      };
      channel.port2.postMessage("");
    } else {
      setTimeout(resolve, 0);
    }
  });
}

/**
 * Delay helper for longer waits between test runs
 */
function delay(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

// ============ Test Scenarios ============

/**
 * Basic single-call scenario
 * Measures time from schedule to execution for a single callback
 */
async function measureBasicOnce(timeoutSeconds: number): Promise<number> {
  const timeoutMs = timeoutSeconds * 1000;
  
  // Yield to prevent nested timer effects
  await yieldToEventLoop();
  
  return new Promise((resolve) => {
    const startTime = now();
    scheduleCallback(() => {
      const elapsed = now() - startTime;
      resolve(elapsed);
    }, timeoutMs);
  });
}

/**
 * Nested scenario - NEST_DEPTH consecutive calls to trigger 4ms clamping
 * Browser rule: "If nesting level is greater than 5, and timeout < 4ms, set timeout to 4ms"
 * 
 * Call sequence:
 * - Call 0: nestingLevel = 0
 * - Call 1: nestingLevel = 1
 * - Call 2: nestingLevel = 2
 * - Call 3: nestingLevel = 3
 * - Call 4: nestingLevel = 4
 * - Call 5: nestingLevel = 5 (4ms clamping starts here!)
 */
async function measureNested(timeoutSeconds: number): Promise<number> {
  const timeoutMs = timeoutSeconds * 1000;
  
  // Yield to start fresh
  await yieldToEventLoop();
  
  return new Promise((resolve) => {
    const startTime = now();
    let callCount = 0;
    
    function nestedCallback() {
      callCount++;
      if (callCount >= NEST_DEPTH) {
        // Finished all nested calls
        const elapsed = now() - startTime;
        resolve(elapsed);
      } else {
        // Schedule next nested call
        scheduleCallback(nestedCallback, timeoutMs);
      }
    }
    
    // Start the chain
    scheduleCallback(nestedCallback, timeoutMs);
  });
}

/**
 * High-frequency scenario (pygame-like)
 * Schedule many callbacks at once and measure until all complete
 */
async function measureHighFrequency(timeoutSeconds: number): Promise<number> {
  const timeoutMs = timeoutSeconds * 1000;
  
  // Yield to start fresh
  await yieldToEventLoop();
  
  return new Promise((resolve) => {
    const startTime = now();
    let completedCount = 0;
    
    // Schedule all callbacks at once
    for (let i = 0; i < BURST_SIZE; i++) {
      scheduleCallback(() => {
        completedCount++;
        if (completedCount === BURST_SIZE) {
          const elapsed = now() - startTime;
          resolve(elapsed);
        }
      }, timeoutMs);
    }
  });
}

/**
 * Game loop scenario (REAL pygame-like)
 * Sequential execution like: while True: do_work(); await sleep(1/fps)
 */
async function measureGameLoop(timeoutSeconds: number): Promise<number> {
  const timeoutMs = timeoutSeconds * 1000;
  
  // Yield to start fresh
  await yieldToEventLoop();
  
  const startTime = now();
  
  // Run BURST_SIZE frames sequentially
  for (let frame = 0; frame < BURST_SIZE; frame++) {
    await new Promise<void>((resolve) => {
      scheduleCallback(() => {
        // Simulate game work (do_something + draw_canvas)
        // In real game, this would be actual game logic
        resolve();
      }, timeoutMs);
    });
  }
  
  const elapsed = now() - startTime;
  return elapsed;
}

// ============ Scenario Runners ============

async function runScenario(
  name: string,
  measureFn: (timeout: number) => Promise<number>,
  reuseMessageChannel: boolean,
  iterations: number = ITERATIONS
): Promise<number[][]> {
  console.log(`\nRunning: ${name}...`);
  const results: number[][] = [];
  
  for (const threshold of THRESHOLDS) {
    const row: number[] = [];
    
    for (const sleepSeconds of SLEEP_VALUES) {
      // Configure scheduler
      setTestSchedulerOptions({ threshold, reuseMessageChannel });
      resetSchedulerState();
      
      // Warm up
      await measureFn(sleepSeconds);
      await delay(10);
      
      // Run measurements
      let totalTime = 0;
      for (let i = 0; i < iterations; i++) {
        const elapsed = await measureFn(sleepSeconds);
        totalTime += elapsed;
        
        // Small delay between iterations to prevent browser slowdown
        if (i % 20 === 19) {
          await delay(5);
        }
      }
      
      const avgTime = totalTime / iterations;
      row.push(avgTime);
    }
    
    results.push(row);
    
    // Delay between thresholds
    await delay(20);
  }
  
  return results;
}

// ============ Results Display ============

function printTable(title: string, data: number[][]) {
  console.log(`\n## ${title}`);
  console.log(`| timeout<=k \\ sleep(s) | ${SLEEP_VALUES.map(s => s.toFixed(3)).join(" | ")} |`);
  console.log(`|${"-".repeat(23)}|${SLEEP_VALUES.map(() => "-".repeat(7)).join("|")}|`);
  
  THRESHOLDS.forEach((threshold, idx) => {
    const row = data[idx];
    const rowStr = row.map(v => v.toFixed(2).padStart(5)).join(" | ");
    console.log(`| ${String(threshold).padEnd(21)} | ${rowStr} |`);
  });
}

function printComparison(normalData: number[][], reuseData: number[][]) {
  console.log(`\n### Performance Comparison (reuse - normal, negative = reuse faster)`);
  console.log(`| timeout<=k \\ sleep(s) | ${SLEEP_VALUES.map(s => s.toFixed(3)).join(" | ")} |`);
  console.log(`|${"-".repeat(23)}|${SLEEP_VALUES.map(() => "-".repeat(7)).join("|")}|`);
  
  THRESHOLDS.forEach((threshold, idx) => {
    const normalRow = normalData[idx];
    const reuseRow = reuseData[idx];
    const diffRow = normalRow.map((v, i) => reuseRow[i] - v);
    const rowStr = diffRow.map(v => {
      const sign = v >= 0 ? '+' : '';
      return (sign + v.toFixed(2)).padStart(5);
    }).join(" | ");
    console.log(`| ${String(threshold).padEnd(21)} | ${rowStr} |`);
  });
}

// ============ Main Test Runner ============

async function main() {
  console.log("=".repeat(60));
  console.log("Scheduler Benchmark Test");
  console.log("=".repeat(60));
  
  // Improved runtime detection
  const g = globalThis as any;
  let runtimeName = "Unknown";
  let runtimeDetails = "";
  
  if (typeof window !== "undefined" && typeof document !== "undefined") {
    // Browser
    runtimeName = "Browser";
    if (typeof navigator !== "undefined") {
      const ua = navigator.userAgent;
      if (/chrome/i.test(ua) && !/edg/i.test(ua)) {
        runtimeDetails = " (Chrome)";
      } else if (/firefox/i.test(ua)) {
        runtimeDetails = " (Firefox)";
      } else if (/safari/i.test(ua) && !/chrome/i.test(ua)) {
        runtimeDetails = " (Safari)";
      } else if (/edg/i.test(ua)) {
        runtimeDetails = " (Edge)";
      }
    }
  } else if (g.Deno?.version?.deno) {
    runtimeName = "Deno";
    runtimeDetails = ` ${g.Deno.version.deno}`;
  } else if (g.process?.versions?.node) {
    runtimeName = "Node.js";
    runtimeDetails = ` ${g.process.versions.node}`;
  }
  
  console.log(`Runtime: ${runtimeName}${runtimeDetails}`);
  console.log(`Iterations per test: ${ITERATIONS}`);
  console.log(`Sleep values (seconds): ${SLEEP_VALUES.join(", ")}`);
  console.log(`Threshold values: ${THRESHOLDS.join(", ")}`);
  console.log(`Nested depth: ${NEST_DEPTH}`);
  console.log(`Burst size: ${BURST_SIZE}`);
  console.log("=".repeat(60));

  // 1. Basic scenario
  const basicNormal = await runScenario(
    "Basic (normal)",
    measureBasicOnce,
    false,
    ITERATIONS
  );
  printTable("Basic execution", basicNormal);

  await delay(100);

  const basicReuse = await runScenario(
    "Basic (MessageChannel reuse)",
    measureBasicOnce,
    true,
    ITERATIONS
  );
  printTable("Basic execution (MessageChannel reuse)", basicReuse);
  printComparison(basicNormal, basicReuse);

  await delay(100);

  // 2. Nested scenario
  const nestedIterations = Math.max(20, Math.floor(ITERATIONS / 5));
  const nestedNormal = await runScenario(
    `Nested (${NEST_DEPTH} deep)`,
    measureNested,
    false,
    nestedIterations
  );
  printTable(`Nested execution (${NEST_DEPTH} deep)`, nestedNormal);

  await delay(100);

  const nestedReuse = await runScenario(
    `Nested (${NEST_DEPTH} deep, MC reuse)`,
    measureNested,
    true,
    nestedIterations
  );
  printTable(`Nested execution (${NEST_DEPTH} deep, MessageChannel reuse)`, nestedReuse);
  printComparison(nestedNormal, nestedReuse);

  await delay(100);

  // 3. High-frequency scenario
  const hfIterations = Math.max(20, Math.floor(ITERATIONS / 10));
  const highFreqNormal = await runScenario(
    "High-frequency (pygame-like)",
    measureHighFrequency,
    false,
    hfIterations
  );
  printTable("High-frequency (pygame-like)", highFreqNormal);

  await delay(100);

  const highFreqReuse = await runScenario(
    "High-frequency (MC reuse)",
    measureHighFrequency,
    true,
    hfIterations
  );
  printTable("High-frequency (MessageChannel reuse)", highFreqReuse);
  printComparison(highFreqNormal, highFreqReuse);

  await delay(100);

  // 4. Game loop scenario
  const gameLoopIterations = Math.max(5, Math.floor(ITERATIONS / 20));
  const gameLoopNormal = await runScenario(
    "Game loop (sequential)",
    measureGameLoop,
    false,
    gameLoopIterations
  );
  printTable("Game loop (sequential)", gameLoopNormal);

  await delay(100);

  const gameLoopReuse = await runScenario(
    "Game loop (MC reuse)",
    measureGameLoop,
    true,
    gameLoopIterations
  );
  printTable("Game loop (sequential, MessageChannel reuse)", gameLoopReuse);
  printComparison(gameLoopNormal, gameLoopReuse);

  console.log("\n" + "=".repeat(60));
  console.log("Tests completed!");
  console.log("=".repeat(60));
  
  // Analysis guide
  console.log("\n## Results Analysis Guide");
  console.log("1. Values show average execution time in milliseconds.");
  console.log("2. Higher timeout<=k means more cases use immediate path.");
  console.log("3. For sleep(0): lower is better (faster scheduling).");
  console.log("4. For sleep(0.001), sleep(0.002): closer to expected timeout is better.");
  console.log("5. Nested tests demonstrate browser's 4ms clamping effect.");
  console.log("6. In comparison tables: negative = reuse faster, positive = normal faster.");
}

// Run tests
main().catch((err) => {
  console.error("Error running benchmark:", err);
  const g = globalThis as any;
  if (typeof g.process !== "undefined") {
    g.process.exit(1);
  }
});

File: test_bench.py
"""
USAGE IN PYODIDE CONSOLE:
--------------------------
Method - Direct execution:
    from pyodide.http import pyfetch
    response = await pyfetch('https://raw.githubusercontent.com/leehanjeong/pyodide-scheduler-benchmark/main/test_bench.py')
    code = await response.string()
    exec(code)
"""

import asyncio
import sys
import time

# Test parameters
SLEEP_VALUES = [0, 0.001, 0.002, 0.016] 
ITERATIONS = 100
BURST_SIZE = 100

def now():
    """Get current time in milliseconds"""
    return time.perf_counter() * 1000


async def measure_basic_once(sleep_sec):
    """
    Scenario 1: Basic single call
    Measures one asyncio.sleep() execution time
    
    Note: await asyncio.sleep(0) resets browser's timer nesting level.
    Browser enforces 4ms minimum when nestingLevel >= 5.
    """
    await asyncio.sleep(0)
    
    start = now()
    await asyncio.sleep(sleep_sec)
    return now() - start


async def measure_high_frequency(sleep_sec):
    """
    Scenario 2: High-frequency concurrent calls
    Like pygame with many sprite updates at once
    
    Schedules 100 asyncio.sleep() calls simultaneously
    """
    await asyncio.sleep(0)
    
    start = now()
    await asyncio.gather(*[
        asyncio.sleep(sleep_sec)
        for _ in range(BURST_SIZE)
    ])
    return now() - start


async def measure_game_loop(sleep_sec):
    """
    Scenario 3: Sequential game loop
    Real pygame pattern: while True: update(); draw(); await sleep(fps)
    
    Runs 100 frames sequentially
    """
    await asyncio.sleep(0)
    
    start = now()
    for _ in range(BURST_SIZE):
        await asyncio.sleep(sleep_sec)
    return now() - start


async def run_scenario(name, measure_fn, iterations=ITERATIONS):
    """Run one scenario for all sleep values"""
    print(f"\n{'='*60}")
    print(f"  {name}")
    print('='*60)
    
    results = []
    
    for sleep_sec in SLEEP_VALUES:
        sleep_ms = sleep_sec * 1000
        
        # Warm up
        await measure_fn(sleep_sec)
        await asyncio.sleep(0.01)
        
        # Run iterations
        times = []
        for i in range(iterations):
            elapsed = await measure_fn(sleep_sec)
            times.append(elapsed)
        
        avg = sum(times) / len(times)
        min_time = min(times)
        max_time = max(times)
        results.append({
            'sleep_ms': sleep_ms,
            'avg': avg,
            'min': min_time,
            'max': max_time,
        })
        
        print(f"  sleep({sleep_ms:>5.1f}ms): avg={avg:>7.2f}ms  min={min_time:>7.2f}ms  max={max_time:>7.2f}ms")
    
    return results


def print_summary_table(all_results):
    """Print comparison table of all scenarios"""
    print("\n" + "="*80)
    print("SUMMARY TABLE (average times in milliseconds)")
    print("="*80)
    
    # Header
    header = f"{'Scenario':<25} | " + " | ".join(f"{s*1000:>6.0f}ms" for s in SLEEP_VALUES)
    print(header)
    print("-"*80)
    
    # Rows
    for scenario_name, results in all_results.items():
        values = " | ".join(f"{r['avg']:>7.2f}" for r in results)
        print(f"{scenario_name:<25} | {values}")
    
    print("="*80)


async def main():
    print("="*80)
    print("SCHEDULER BENCHMARK TEST")
    print("="*80)
    
    # Runtime detection 
    runtime = "Unknown"
    try:
        from js import navigator
        ua = str(navigator.userAgent)
        if "Chrome" in ua and "Edg" not in ua:
            runtime = "Pyodide/Chrome"
        elif "Firefox" in ua:
            runtime = "Pyodide/Firefox"
        elif "Safari" in ua and "Chrome" not in ua:
            runtime = "Pyodide/Safari"
        elif "Edg" in ua:
            runtime = "Pyodide/Edge"
        else:
            runtime = "Pyodide/Browser"
    except:
        runtime = f"Python {sys.version.split()[0]}"
    
    print(f"Runtime: {runtime}")
    print(f"Sleep values: {', '.join(f'{s*1000:.0f}ms' for s in SLEEP_VALUES)}")
    print(f"Iterations: {ITERATIONS}")
    print(f"Burst size: {BURST_SIZE}")
    print("="*80)
    
    all_results = {}
    
    # Scenario 1: Basic
    print("\n🔹 Running Scenario 1...")
    basic_results = await run_scenario(
        "1. Basic (single call)",
        measure_basic_once,
        ITERATIONS
    )
    all_results["1. Basic"] = basic_results
    
    await asyncio.sleep(0.1)
    
    # Scenario 2: High-frequency (fewer iterations)
    print("\n🔹 Running Scenario 2...")
    hf_iters = max(20, ITERATIONS // 10)
    hf_results = await run_scenario(
        f"2. High-freq ({BURST_SIZE}x)",
        measure_high_frequency,
        hf_iters
    )
    all_results[f"2. High-freq ({BURST_SIZE}x)"] = hf_results
    
    await asyncio.sleep(0.1)
    
    # Scenario 3: Game loop (fewer iterations)
    print("\n🔹 Running Scenario 3...")
    gl_iters = max(5, ITERATIONS // 20)
    gl_results = await run_scenario(
        f"3. Game loop ({BURST_SIZE}x)",
        measure_game_loop,
        gl_iters
    )
    all_results[f"3. Game loop ({BURST_SIZE}x)"] = gl_results
    
    # Print summary
    print_summary_table(all_results)
    


# Entry point
if __name__ == "__main__":
    asyncio.run(main())
